### PR TITLE
feat(categories): add new endpoint to list realtime host / service categories

### PR DESCRIPTION
**Pull request by TamazC**
## Description

This adds a new endpoint to list the host categories from the real time

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

<h2> How this pull request can be tested ? </h2>

will be tested with the overall feature

## Checklist

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Core/Infrastructure/RealTime/Repository/Tag/DbTagFactory.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+{
+    /**
+     * Create ServiceCategory model using data from database
+     *
+     * @param array<string, mixed> $data
+     * @return Tag
+     */
+    public static function createFromRecord(array $data): Tag
+    {
+        return new Tag((int) $data['id'], $data['name'], $data['type']);
```
Comment (kduret) [inline src/Core/Infrastructure/RealTime/Repository/Tag/DbTagFactory.php:17]:
```suggestion
        return new Tag($data['id'], $data['name'], $data['type']);
```

File: src/Core/Infrastructure/RealTime/Api/FindHostCategory/FindHostCategoryController.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+declare(strict_types=1);
+
+namespace Core\Infrastructure\RealTime\Api\FindHostCategory;
+
+use Centreon\Application\Controller\AbstractController;
+use Core\Application\RealTime\UseCase\FindHostCategory\FindHostCategory;
+
+class FindHostCategoryController extends AbstractController
+{
+    public function __invoke(FindHostCategory $useCase, FindHostCategoryPresenter $presenter): object
```
[the remark this answers is not in the document]
Comment (adr-mo) [reply src/Core/Infrastructure/RealTime/Api/FindHostCategory/FindHostCategoryController.php:30]: Done ✅

File: src/Core/Infrastructure/RealTime/Repository/Tag/DbTagFactory.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?php
+
```
[the remark this answers is not in the document]
Comment (adr-mo) [reply src/Core/Infrastructure/RealTime/Repository/Tag/DbTagFactory.php]: Done ✅

File: src/Core/Infrastructure/RealTime/Repository/Tag/DbTagFactory.php (hunk 1 above)
Comment (adr-mo) [reply src/Core/Infrastructure/RealTime/Repository/Tag/DbTagFactory.php]: Done ✅ (we've re-added the casting for both int)

File: tests/php/Core/Application/RealTime/UseCase/FindServiceCategory/FindServiceCategoryTest.php [hunk 4]
```diff
@@ -22,24 +22,24 @@
...
 use Tests\Core\Application\RealTime\UseCase\FindServiceCategory\FindServiceCategoryPresenterStub;
 
 it('Find all service categories', function () {
-    $category = new ServiceCategory(1, 'category-name');
-    $repository = $this->createMock(ReadServiceCategoryRepositoryInterface::class);
+    $category = new Tag(1, 'service-category-name', Tag::SERVICE_CATEGORY_TYPE_ID);
+    $repository = $this->createMock(ReadTagRepositoryInterface::class);
     $repository->expects($this->once())
-        ->method('findAll')
+        ->method('findAllByType')
```
Comment (kduret) [inline tests/php/Core/Application/RealTime/UseCase/FindServiceCategory/FindServiceCategoryTest.php:34]:
```suggestion
        ->method('findAllByTypeId')
```

File: tests/php/Core/Application/RealTime/UseCase/FindHostCategory/FindHostCategoryTest.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+use Core\Domain\RealTime\Model\Tag;
+use Core\Application\RealTime\Repository\ReadTagRepositoryInterface;
+use Core\Application\RealTime\UseCase\FindHostCategory\FindHostCategory;
+use Tests\Core\Application\RealTime\UseCase\FindHostCategory\FindHostCategoryPresenterStub;
+
+it('Find all host categories', function () {
+    $category = new Tag(1, 'host-category-name', Tag::HOST_CATEGORY_TYPE_ID);
+    $repository = $this->createMock(ReadTagRepositoryInterface::class);
+    $repository->expects($this->once())
+        ->method('findAllByType')
```
Comment (kduret) [inline tests/php/Core/Application/RealTime/UseCase/FindHostCategory/FindHostCategoryTest.php:34]:
```suggestion
        ->method('findAllByTypeId')
```

**Review by kduret — APPROVED** [review]

**Review by callapa — CHANGES_REQUESTED** [review]

File: src/Core/Application/RealTime/UseCase/FindHostCategory/FindHostCategory.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+     * Checks if at least on monitoring server has BBDO protocol in version 3.0.0
+     *
+     * @return boolean
+     */
+    private function isBBDOVersionCompatible(): bool
+    {
+        $brokerConfigurations = $this->brokerRepository->findAllByParameterName(self::BBDO_VERSION_CONFIG_KEY);
+        foreach ($brokerConfigurations as $brokerConfiguration) {
+            if (
+                version_compare(
```
Comment (callapa) [inline src/Core/Application/RealTime/UseCase/FindHostCategory/FindHostCategory.php:111]: No check on the 'daemon' property ?

File: src/Core/Application/RealTime/UseCase/FindServiceCategory/FindServiceCategory.php [hunk 7]
```diff
@@ -60,11 +79,44 @@ public function __invoke(FindServiceCategoryPresenterInterface $presenter): void
...
+        $this->error($message);
+        $presenter->setResponseStatus(new ImcompatibilityResponse($message));
+    }
+
+    /**
+     * Checks if at least on monitoring server has BBDO protocol in version 3.0.0
+     *
+     * @return boolean
+     */
+    private function isBBDOVersionCompatible(): bool
```
Comment (callapa) [inline src/Core/Application/RealTime/UseCase/FindServiceCategory/FindServiceCategory.php:107]: If you want you can use a trait to share this part of code with the other use case **FindHostCategory**

File: src/Core/Domain/RealTime/Model/Tag.php [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+     *
+     * @param int $typeId
+     * @return void
+     * @throws \InvalidArgumentException
+     */
+    private function validateTypeId(int $typeId): void
+    {
+        if (! in_array($typeId, self::getAvailableTypeIds())) {
+            throw new \InvalidArgumentException('Type Id is not valid');
+        }
```
Comment (callapa) [inline src/Core/Domain/RealTime/Model/Tag.php:95]: Move it directly into the setter

File: tests/php/Core/Application/RealTime/UseCase/FindHostCategory/FindHostCategoryPresenterStub.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    public $response;
+
+    /**
+     * @var ResponseStatusInterface|null
+     */
+    protected $responseStatus;
+
+    /**
+     * constructor
+     */
```
Comment (callapa) [inline tests/php/Core/Application/RealTime/UseCase/FindHostCategory/FindHostCategoryPresenterStub.php:45]: Not needed for this one ;)

File: src/Core/Application/RealTime/UseCase/FindHostCategory/FindHostCategory.php (hunk 6 above)
Comment (adr-mo) [reply src/Core/Application/RealTime/UseCase/FindHostCategory/FindHostCategory.php:111]: I guess not, as the method `findAllByParameterName` returns only a `BrokerConfiguration` entity which represent only a subset of the configuration (one key and one value).

To get this parameter, an other call to `findAllByParameterName` needs to be done. Do we need to check for the daemon property ? IDK

File: src/Core/Application/Common/Broker/BrokerTrait.php [hunk 10]
```diff
@@ -0,0 +1,49 @@
+<?php
```
[the remark this answers is not in the document]
Comment (adr-mo) [reply src/Core/Application/Common/Broker/BrokerTrait.php:1]: Move this to an other class BBDOBroker.php (abstract or not to discuss)

File: doc/API/centreon-api-v22.04.yaml [hunk 11]
```diff
@@ -294,6 +295,31 @@ paths:
           $ref: '#/components/responses/Forbidden'
         '500':
           $ref: '#/components/responses/InternalServerError'
+  /monitoring/services/categories:
```
Comment (callapa) [inline doc/API/centreon-api-v22.04.yaml:298]: You forgot to document the API endpoint monitoring/hosts/categories.

**Review by callapa — CHANGES_REQUESTED** [review]

**Review by callapa — CHANGES_REQUESTED** [review]

**Review by callapa — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Pull request merged**